plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,398 @@
+we made are:
+
+- No alpha equivalence check, i.e., `\x -> x` and `\y -> y` are considered different expressions.
+- The builtin function arity information is approximate: rather than using the accurate arities,
+  we simply use the maximum number of arguments applied to a builtin function in the program
+  as the builtin function's arity. The arity information is used to determine whether a builtin
+  application is possibly saturated.
+
+-------------------------------------------------------------------------------
+2. How does it work?
+-------------------------------------------------------------------------------
+
+We use the following example to explain how the implementation works:
+
+\x y -> (1+(2+x))
+        +
+        (case y [ (1+(2+x)) + (3+x)
+                , (2+x) + (3+x)
+                , 4+x
+                ]
+        )
+
+The implementation makes several passes on the given term. The first pass collects builtin
+arity information as described above.
+
+In the second pass, we assign a unique ID to each `LamAbs`, `Delay`, and each `Case` branch.
+Then, we annotate each subterm with a path, consisting of IDs encountered from the root
+to that subterm (not including itself).
+
+In the above example, the ID of `\x` is 0, the ID of `\y` is 1, and the IDs of the
+three case branches are 2, 3, 4 (the actual numbers don't matter, as long as they are unique).
+The path for the first `1+(2+x)` and the first `2+x` is "0.1"; the path for the second
+`1+(2+x)` and the second `2+x` is "0.1.2"; the path for `4+x` is "0.1.4".
+
+In the third pass, we calculate a count for each `(term, path)` pair, where `term` is a
+non-workfree term, and `path` is its path. If the same term has two paths, and one is an
+ancestor (i.e., prefix) of the other, we increment the count for the ancestor path in both
+instances.
+
+In the above example, there are three occurrences of `2+x`, whose paths are "0.1", "0.1.2"
+and "0.1.3", respectively. The first path is an ancestor of the latter two. Therefore,
+the count for `(2+x, "0.1")` is 3, while the count for `(2+x, "0.1.2")` and `(2+x, "0.1.3")`
+is 0. The following all have a count of 1: `(3+x, "0.1.2")`, `(3+x, "0.1.3")` and
+`(4+x, "0.1.4")`.
+
+Now, each `(term, path)` pair whose count is greater than 1 is a CSE candidate.
+In the above example, the CSE candidates are `(2+x, "0.1")` and `(1+(2+x), "0.1")`.
+Note that `3+x` is not a CSE candidate, because it has two paths, and neither has a count
+greater than 1. `2+` is also not a CSE candidate, because it is workfree.
+
+The CSE candidates are then processed in descending order of their `termSize`s. For each CSE
+candidate, we generate a fresh variable, create a LamAbs for it under its path, and substitute
+it for all occurrences in the original term whose paths are descendents (or self) of
+the candidate's path. The order is because a bigger expression may contain a small subexpression.
```

Comment:
No because the bigger CS contains the smaller CS.
You can do it the other way around, but then you not only need to perform substitution in the original term, but also in the CS themselves

Reply:
> But you do perform substitution in CSs?

No, what I meant to say: if we process `1+(2+x)` first, then we only need to perform substitution in the original term itself. On the other hand, if we process `2+x` first, we must also perform substitution in the CSs, i.e., turning `1+(2+x)` into `1+cse2`. Otherwise, when we process `1+(2+x)` later, we will not find any `2+x` in the original term.

Reply:
To be clear, when you say "perform substitution in the original term itself", you mean "the partially processed term with some candidates hoisted out"? Because you do want to hoist out the instance of `(2+x)` in the RHS of the newly created binding for the first CSE candidate, right?

Reply:
> you mean "the partially processed term with some candidates hoisted out"?

Yes.